cachi2/core/models/output.py

```diff
@@ -43,12 +44,53 @@ def resolve_value(self, mappings: Dict[str, str]) -> str:
         substitute all placeholders defined; if no placeholders are contained within the value
         string, substitution is a NOOP (e.g. legacy "literal" variables)
         """
+
+        def get_placeholders(t: string.Template) -> Set[str]:
+            """Return a set of placeholders in a template.
+
+            Implementation is based on [1] which appeared in 3.11 [2] without additional error
+            handling which we don't need for our mostly internal use case.
+            [1] https://github.com/python/cpython/blob/3b4cd48d2988e74405838accde5edcc3b71bec48/Lib/string.py#L157
+            [2] https://docs.python.org/3.11/library/string.html#string.Template.get_identifiers
+
+            TODO: Drop this after bumping minimum required version to 3.11
+            """
+            placeholders = set()
+            matches = t.pattern.finditer(t.template)
+            for m in matches:
+                if placeholder := m.group("named") or m.group("braced"):
+                    placeholders.add(placeholder)
+
+            return placeholders
+
         # legacy path variable handling, need to prepend the base path placeholder
         if self.kind == "path" and "output_dir" in mappings:
             log.debug(f"Adjusting a legacy path variable value '{self.name}:{self.value}'")
             self.value = "${output_dir}/" + self.value
 
-        return string.Template(self.value).safe_substitute(mappings)
+        # "Recursively" resolve potentially nested variables up to len(mappings) tries
+        ret = self.value
+        substituted: Set[str] = set()
+        for i, m in enumerate(mappings):
+            log.debug(f"EnvironmentVariable resolution iteration {i+1}.: {ret}")
+
+            t_old = string.Template(ret)
+            ret = t_old.safe_substitute(mappings)
+            t_new = string.Template(ret)
+            p_new = get_placeholders(t_new)
+
+            if substituted & p_new:
+                raise Cachi2Error(
+                    f"Detected a cycle in environment variable expansion of '{self.name}'",
+                    solution=(
+                        "Inspect all relevant environment variables and make sure their "
+                        "expansion doesn't lead to a cycle. Ideally, avoiding nesting of "
+                        "variables altogether."
+                    ),
+                )
+            substituted |= get_placeholders(t_old) - p_new
```

Comment:
If I understand correctly, this is needed for cases such as:

```python
before_substitution = "${known_variable} and ${unknown_variable}"
after_substitution = "some_value and ${unknown_variable}"
```
We don't want the `unknown_variable` to cause false-positive cycle detection errors, so we avoid marking it as substituted

I wonder if this can lead to false negatives as well though
```python
before_substitution = "${foo} and ${bar}"
after_substitution = "${bar} and ${foo}"
# get_placeholders(before_substitution) - get_placeholders(after_substitution) == empty set ?
```

Reply:
> I wonder if this can lead to false negatives as well though
>```
>before_substitution = "${foo} and ${bar}"
>after_substitution = "${bar} and ${foo}">
># get_placeholders(before_substitution) - get_placeholders(after_substitution) == empty set
>```

@chmeliik if we add an additional `before_substitution != after_substitution` that would take care of the problem wouldn't it?

Reply:
Hmm, not sure. The cycles can be more complicated
```python
mappings = {
    "cycle_part1": "$cycle_part3",
    "cycle_part2": "$cycle_part1",
    "cycle_part3": "$cycle_part2"
}
```
```python
"$cycle_part1 $cycle_part2"
# ->
"$cycle_part3 $cycle_part1"  # substituted = {"cycle_part2"}
# ->
"$cycle_part2 $cycle_part3"  # substituted = {"cycle_part2", "cycle_part1"}
# cycle detected
```
In this case I believe the cycle will be caught after the second iteration (even though it could be caught after the first one), but I'm not sure if every cycle will be caught

Reply:
Maybe, rather than subtracting the new placeholders from the `substituted` set, a more robust solution would be to intersect `substituted` with the known `mappings`?

```python
substituted |= get_placeholders(t_old) & mappings.keys()
```

That should still cover the "don't mark unknown variables as substituted" case and shouldn't interfere with cycle detection for known variables

Reply:
> Hmm, not sure. The cycles can be more complicated
> 
> ```python
> mappings = {
>     "cycle_part1": "$cycle_part3",
>     "cycle_part2": "$cycle_part1",
>     "cycle_part3": "$cycle_part2"
> }
> ```
> 
> ```python
> "$cycle_part1 $cycle_part2"
> # ->
> "$cycle_part3 $cycle_part1"  # substituted = {"cycle_part2"}
> # ->
> "$cycle_part2 $cycle_part3"  # substituted = {"cycle_part2", "cycle_part1"}
> # cycle detected
> ```
> 
> In this case I believe the cycle will be caught after the second iteration (even though it could be caught after the first one), but I'm not sure if every cycle will be caught

I guess there is an algorithm that could catch it after the first iteration, but as long as it's detected it's not an issue.

Reply:
> Maybe, rather than subtracting the new placeholders from the `substituted` set, a more robust solution would be to intersect `substituted` with the known `mappings`?
> 
> ```python
> substituted |= get_placeholders(t_old) & mappings.keys()
> ```
> 
> That should still cover the "don't mark unknown variables as substituted" case and shouldn't interfere with cycle detection for known variables

It felt a bit counter intuitive, but yeah, this seems to work, maybe I just rename it to `seen` rather than substituted that actually might describe it better.

Reply:
IMO `substituted` is still more accurate. In each iteration, every placeholder in the template string really does get substituted - unless it has no mapping in `mappings`